### PR TITLE
feat: allow passing base_url in openai section of config.toml

### PR DESCRIPTION
**Pull request by maciejmajek**
## Purpose

Some local model vendors support OpenAI api.

## Proposed Changes

This Pr introduces changes that allow user to specify 
- base_url
in the openai section of config.toml to use the langchain's OpenAI abstractions in ollama/vllm and so on.

## Issues

- Links to relevant issues

## Testing

config.toml
```toml
[vendor]
simple_model = "openai"
complex_model = "openai"
embeddings_model = "openai"
...
[openai]
simple_model = "gpt-4o-mini"
complex_model = "qwen2.5-coder:3b"
embeddings_model = "text-embedding-ada-002"
base_url = "http://localhost:11434/v1/" # for openai compatible apis
...
```

```python
04:50:10 (rai-b2azRMqA-py3.12) mmajek@robo-pc-054 rai ±|feat/openai-compatible-apis ✗|→ python
Python 3.12.3 (main, Nov  6 2024, 18:32:19) [GCC 13.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from rai.utils.model_initialization import get_llm_model 
>>> llm = get_llm_model("complex_model")
2024-11-29 16:50:22 robo-pc-054 rai.utils.model_initialization[461106] INFO Using LLM model: openai-qwen2.5-coder:3b
>>> llm.invoke("hello")
2024-11-29 16:50:29 robo-pc-054 httpx[461106] INFO HTTP Request: POST http://localhost:11434/v1/chat/completions "HTTP/1.1 200 OK"
AIMessage(content='Hello! How can I assist you today?', additional_kwargs={'refusal': None}, response_metadata={'token_usage': {'completion_tokens': 10, 'prompt_tokens': 30, 'total_tokens': 40, 'completion_tokens_details': None, 'prompt_tokens_details': None}, 'model_name': 'qwen2.5-coder:3b', 'system_fingerprint': 'fp_ollama', 'finish_reason': 'stop', 'logprobs': None}, id='run-d5c590c4-56ed-4174-91f0-f0d78844570f-0', usage_metadata={'input_tokens': 30, 'output_tokens': 10, 'total_tokens': 40, 'input_token_details': {}, 'output_token_details': {}})
```

## Summary by CodeRabbit

- **New Features**
	- Enhanced OpenAI configuration support with new parameters: `base_url`.
	- Introduced a dedicated configuration class for OpenAI settings, improving error handling for API key management.

- **Bug Fixes**
	- Improved error handling when the `api_key` is not set, ensuring better user feedback.

### CODE REVIEW

**Comment by maciejmajek** [comment]
@coderabbit full review

File: src/rai/rai/utils/model_initialization.py [hunk 1]
```diff
@@ -110,17 +117,31 @@ def get_llm_model(
...
+            model_config.api_key
+            if model_config.api_key != ""
+            else os.getenv("OPENAI_API_KEY", None)
+        )
+        if api_key is None:
+            raise ValueError("OPENAI_API_KEY is not set")
+
+        return ChatOpenAI(
+            model=model, base_url=model_config.base_url, api_key=SecretStr(api_key)
+        )
```
[the remark this answers is not in the document]
Comment (maciejmajek) [reply src/rai/rai/utils/model_initialization.py]: Fixed

File: config.toml [hunk 2]
```diff
@@ -13,6 +13,8 @@ region_name = "us-east-1"
 simple_model = "gpt-4o-mini"
 complex_model = "gpt-4o-2024-08-06"
 embeddings_model = "text-embedding-ada-002"
+base_url = "https://api.openai.com/v1/" # for openai compatible apis
+api_key = "" # for openai compatible apis, overrides the environment variable OPENAI_API_KEY
```
Comment (boczekbartek) [inline config.toml]: I can't understand:
- if `api_key = ""` - will OPENAI_API_KEY be overwritten by ""? 
- the main purpose is to support local models with compatible api to openai, right? But then why do they require `api_key` if they are local?
- I'm afraid that storing `api_key` in a file that is under git will eventually result in leaking the key by someone doing `git commit -am "msg"`

**Review by boczekbartek — COMMENTED** [review]
@maciejmajek 
Haven't tested the functionallity yet, but I have some questions. Could you have a look?

File: config.toml (hunk 2 above)
Comment (maciejmajek) [reply config.toml]: "" won't overwrite
The api key is used for eg ollama and in that case the key should be equal to "ollama". Local vendors specify what should be used for api_key in order to properly connect to endpoint.

File: src/rai/rai/utils/model_initialization.py [hunk 3]
```diff
@@ -106,21 +113,38 @@ def get_llm_model(
...
     if vendor == "openai":
         from langchain_openai import ChatOpenAI
 
-        return ChatOpenAI(model=model)
+        model_config = cast(OpenAIConfig, model_config)
+        api_key = (
+            model_config.api_key
+            if model_config.api_key != ""
+            else os.getenv("OPENAI_API_KEY", None)
+        )
```
Comment (boczekbartek) [inline src/rai/rai/utils/model_initialization.py]: Please remove `model_config.api_key`

Comment (maciejmajek) [reply src/rai/rai/utils/model_initialization.py]: Done

**Comment by boczekbartek** [comment]
@maciejmajek 
I cross-checked this feature with [debugging-assistant](https://github.com/RobotecAI/rai/pull/326) and I have some findings:
- with ollama implementation of openai api, tool calling fails (tool is not called,but the string is printed)
- with ollama api tool calling works
- I'll check with vllm and let you know.

**Comment by boczekbartek** [comment]
I tested `vllm serve` with:
```
vllm serve --port 8090 --gpu-memory-utilization 0.99 --max_model_len 10000 --enable-auto-tool-choice --tool-call-parser hermes Qwen/Qwen2.5-7B-Instruct
```
and tool calling worked great through openai API!

**Review by boczekbartek — APPROVED** [review]

**Pull request merged**